### ISSUE TITLE
Performance issue for complex SELECT Queries

### ISSUE THREAD

**Issue by snehalbhavsar916** [yugabyte/yugabyte-db#5865]
Jira Link: [DB-1805](https://yugabyte.atlassian.net/browse/DB-1805)
Yugabyte (YSQL) SELECT Queries performance issues is observer in yugabyte-2.2.2.0, Despite of all the indexes being used, the query is still slower and have comparatively poor performance than PostgreSQL. Here there is a huge difference in the performance between the two.

**Yugabyte (YSQL)**

 (42156 rows)
Time: 407614.142 ms (06:47.614)
```
                                                               QUERY PLAN                                                               
----------------------------------------------------------------------------------------------------------------------------------------
 Nested Loop  (cost=8.25..12.74 rows=1 width=1600)
   ->  Nested Loop  (cost=8.25..12.62 rows=1 width=1088)
         ->  Nested Loop  (cost=8.25..12.51 rows=1 width=1088)
               ->  Nested Loop  (cost=8.25..12.39 rows=1 width=1088)
                     ->  Unique  (cost=8.25..8.27 rows=1 width=1084)
                           ->  Sort  (cost=8.25..8.25 rows=1 width=1084)
                                 Sort Key: n.neid, p.neid, n.geographyl4id_fk, n.cellnum, p.networkelementid_pk, n.vendor
                                 ->  Nested Loop  (cost=0.00..8.24 rows=1 width=1084)
                                       ->  Index Scan using networkelement_pkey on networkelement n  (cost=0.00..4.12 rows=1 width=666)
                                             Filter: ((NOT deleted) AND ((domain)::text = 'RAN'::text))
                                       ->  Index Scan using networkelement_pkey on networkelement p  (cost=0.00..4.11 rows=1 width=422)
                                             Index Cond: (networkelementid_pk = n.networkelementid_fk)
                                             Filter: ((NOT deleted) AND (neid IS NOT NULL))
                     ->  Index Scan using geographyl4_pkey on geographyl4 l4  (cost=0.00..4.11 rows=1 width=8)
                           Index Cond: (geographyl4id_pk = n.geographyl4id_fk)
               ->  Index Scan using geographyl3_pkey on geographyl3 l3  (cost=0.00..0.11 rows=1 width=8)
                     Index Cond: (geographyl3id_pk = l4.geographyl3id_fk)
         ->  Index Scan using geographyl2_pkey on geographyl2 l2  (cost=0.00..0.11 rows=1 width=8)
               Index Cond: (geographyl2id_pk = l3.geographyl2id_fk)
   ->  Index Scan using geographyl1_pkey on geographyl1 l1  (cost=0.00..0.11 rows=1 width=520)
         Index Cond: (geographyl1id_pk = l2.geographyl1id_fk)
(22 rows)

Time: 407614.142 ms (06:47.614)
```
**Where as in PostgreSQL available on same VM:**
```
(42156 rows)
Time: 697.083 ms

I am executing this query in yugabyte and postgresql both:

yb_test_db=# SELECT 
yb_test_db(#         ne.child,
yb_test_db(#             ne.parent,
yb_test_db(#             ne.l4,
yb_test_db(#             ne.cell,
yb_test_db(#             ne.networkelementid_pk,
yb_test_db(#             ne.domain,
yb_test_db(#             ne.vendor,
yb_test_db(#             l1.name l1
yb_test_db(#             
yb_test_db(#     FROM
yb_test_db(#         (SELECT 
yb_test_db(#         n.neid child,
yb_test_db(#             p.neid parent,
yb_test_db(#             n.geographyl4id_fk l4,
yb_test_db(#             n.cellnum cell,
yb_test_db(#             p.networkelementid_pk,
yb_test_db(#             n.domain,
yb_test_db(#             n.vendor
yb_test_db(#     FROM
yb_test_db(#         NetworkElement n
yb_test_db(#     JOIN NetworkElement p ON p.networkelementid_pk = n.networkelementid_fk
yb_test_db(#     WHERE 
yb_test_db(#         n.DELETED = false AND p.DELETED = false
yb_test_db(#             AND n.domain = 'RAN') ne
yb_test_db(#     JOIN GeographyL4 l4 ON l4.geographyl4id_pk = ne.l4
yb_test_db(#     JOIN GeographyL3 l3 ON l3.geographyl3id_pk = l4.geographyl3id_fk
yb_test_db(#     JOIN GeographyL2 l2 ON l2.geographyl2id_pk = l3.geographyl2id_fk
yb_test_db(#     JOIN GeographyL1 l1 ON l1.geographyl1id_pk = l2.geographyl1id_fk  
yb_test_db(# where (ne.parent IS NOT NULL);
```
Also, I am sharing details for yugabyte cluster I am using:

![image](https://user-images.githubusercontent.com/72077131/94568293-f0e13c00-0289-11eb-9db8-4c0494fb47a9.png)

I am not getting what is the issue here? Why the performance is so poor for the select queries in YSQL?

**Comment by ddorian** [comment]
Hi @snehalbhavsar916 

1. Can you share the cpu,ram,ssd of the vm ?
2. Can you try with [colocated database](https://docs.yugabyte.com/latest/api/ysql/commands/ddl_create_database/#colocated) ?
3. It's not fair to have 3 yb-tservers + 3 yb-masters with replication VS 1 postgresql process, no ?

The difference in performance is because YugabyteDB is meant to be used in distributed environments. The query that you made had to bounce multiple times to all 3 processes, adding in serialize/deserialize/cpu/disk/ram overhead.

**Comment by snehalbhavsar916** [comment]
Hi @ddorian 

1. Please find CPU, RAM details as below and the disk is SSD:

[root@stg-ds-fores101zc ~]# lscpu
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                20
On-line CPU(s) list:   0-19
Thread(s) per core:    1
Core(s) per socket:    1
Socket(s):             20
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel Xeon Processor (Skylake, IBRS)
Stepping:              4
CPU MHz:               2399.998
BogoMIPS:              4799.99
Virtualization:        VT-x
Hypervisor vendor:     KVM
Virtualization type:   full
L1d cache:             32K
L1i cache:             32K
L2 cache:              4096K
L3 cache:              16384K
NUMA node0 CPU(s):     0-19
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology eagerfpu pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 arat pku ospke spec_ctrl intel_stibp
[root@stg-ds-fores101zc ~]# 
[root@stg-ds-fores101zc ~]# 
[root@stg-ds-fores101zc ~]# free -h
              total        used        free      shared  buff/cache   available
Mem:            62G         28G        4.0G        150M         30G         33G
Swap:           15G        2.7G         13G
[root@stg-ds-fores101zc ~]# 

2. I have already created colocated database.

3. Lets consider this performance issues is happening due to the current cluster architecture of having rf 3 on single node, ie., having local cluster. But what if the data will be distributed within three different nodes, then also the select query will have to hop on three different servers to fetch the data. In that case there may be possibility to have more slowness issue, since it will also depend on network latency.

I have also checked this queries with having single node yb-ctl cluster, there also we are facing same issue, which is also an ssd server.

**Comment by snehalbhavsar916** [comment]
Perhaps this is not the same as DISTINCT. #5866 is not same as #5865

**Comment by snehalbhavsar916** [comment]
We have created a database in yugabyte with colocated=true. It means that the data is replicated but not distributed, in this case can we compare performance of select Queries with MySQL or PostgreSQL?
Since @ddorian  you have suggested us to make our database colocated above. So we did. Now again the performance of SELECT queries aren't comparable with other databases and is having very poor performance.
